Sources/NIO/ByteBuffer-core.swift

```diff
@@ -800,6 +804,53 @@ extension ByteBuffer {
     }
 }
 
+extension ByteBuffer {
+    /// Errors thrown when calling `copyBytes`.
+    public enum CopyBytesError: Error, Equatable {
+        /// The length of the bytes to copy was negative.
+        case negativeLength
+
+        /// The bytes to copy are not readable.
+        case unreadableSourceBytes
+    }
+
+    /// Copies `length` `bytes` starting at the `fromIndex` to `toIndex`. Does not move the writer index.
+    ///
+    /// - Note: Overlapping ranges, for example `copyBytes(at: 1, to: 2, length: 5)` are allowed.
+    /// - Precondition: The range represented by `fromIndex` and `length` must be readable bytes,
+    ///     that is: `fromIndex >= readerIndex` and `fromIndex + length <= writerIndex`.
+    /// - Parameter fromIndex: The index of the first byte to copy.
+    /// - Parameter toIndex: The index into to which the first byte will be copied.
+    /// - Parameter length: The number of bytes which should be copied.
+    @discardableResult
+    @inlinable
+    public mutating func copyBytes(at fromIndex: Int, to toIndex: Int, length: Int) throws -> Int {
```

Comment:
Throwing errors is a slightly unusual property for `ByteBuffer`: we normally use preconditions or optional returns. I think the only reason this has to throw is because it's public API, otherwise the error cases in which we throw could technically be other things, right?

Are we happy adding this public API to `ByteBuffer` @weissi?

Reply:
I actually wonder if this API should be `internal` for now and we can simply replace all the `throws` with `precondition`. The only callers of this API use `try!` anyway.

Reply:
Discussed both options with @weissi before – IIRC the preference for `public` over `internal` wasn't strong. I'm happy to do either.

Reply:
Let's leave this as-is.